### PR TITLE
add instrumentation for RestTemplateBuilder

### PR DESCRIPTION
**Pull request by zeitlinger**

### CODE REVIEW

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateBeanPostProcessor.java [hunk 1]
```diff
@@ -27,20 +28,22 @@ public Object postProcessAfterInitialization(Object bean, String beanName) {
...
-    RestTemplate restTemplate = (RestTemplate) bean;
-    ClientHttpRequestInterceptor interceptor =
-        SpringWebTelemetry.create(openTelemetryProvider.getObject()).newInterceptor();
-    addRestTemplateInterceptorIfNotPresent(restTemplate, interceptor);
-    return restTemplate;
+    return addRestTemplateInterceptorIfNotPresent(
+        (RestTemplate) bean, openTelemetryProvider.getObject());
   }
 
-  private static void addRestTemplateInterceptorIfNotPresent(
```
Comment (jeanbisutti) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateBeanPostProcessor.java:37]: The modification of this class does not seem related to the instrumentation of the RestTemplateBuilder?

Comment (jeanbisutti) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateBeanPostProcessor.java:37]: Oh, I see now that the method is also used in `SpringWebInstrumentationAutoConfiguration`. Perhaps put the method use by both instrumentation in another class, something like `RestTemplateInterceptor`?

Comment (zeitlinger) [reply instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateBeanPostProcessor.java:37]: good idea - done

**Review by jeanbisutti — COMMENTED** [review]
The `shouldSendTelemetry` test method verifies many things.

Would it be possible to test the `RestTemplateBuilder` instrumentation from another test method calling an HTTP endpoint different from `/ping` (the one called in the `shouldSendTelemetry` test method)?

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+package io.opentelemetry.instrumentation.spring.autoconfigure.instrumentation.web;
+
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.instrumentation.spring.web.v3_1.SpringWebTelemetry;
+import java.util.List;
+import org.springframework.http.client.ClientHttpRequestInterceptor;
+import org.springframework.web.client.RestTemplate;
+
+public class RestTemplateInterceptor {
```
Comment (jeanbisutti) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateInterceptor.java:15]:
```suggestion
class RestTemplateInterceptor {
```

**Comment by zeitlinger** [comment]
> The `shouldSendTelemetry` test method verifies many things.
> 
> Would it be possible to test the `RestTemplateBuilder` instrumentation from another test method calling an HTTP endpoint different from `/ping` (the one called in the `shouldSendTelemetry` test method)?

good idea :smile: - done

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java [hunk 3]
```diff
@@ -157,6 +166,43 @@ AutoConfigurationCustomizerProvider propagatorCustomizer() {
     }
   }
 
+  @BeforeEach
+  void setUp() {
+    if (!initialized) {
+      initialized = true;
+      assertFirstTimeSetup();
```
Comment (jeanbisutti) [inline smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:173]: Having assertions on multiple places makes difficult to know what is verified. Could you please move the assertions contained in the `setUp` method to a test method?

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java [hunk 4]
```diff
@@ -157,7 +168,19 @@ AutoConfigurationCustomizerProvider propagatorCustomizer() {
...
+  }
+
+  private static void reset() {
+    SPAN_EXPORTER.reset();
+    METRIC_EXPORTER.reset();
+    LOG_RECORD_EXPORTER.reset();
+  }
+
   @Test
+  @org.junit.jupiter.api.Order(10)
```
Comment (laurit) [inline smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:183]: any reason this can't be imported?

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+package io.opentelemetry.instrumentation.spring.autoconfigure.instrumentation.web;
+
+import com.google.errorprone.annotations.CanIgnoreReturnValue;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.instrumentation.spring.web.v3_1.SpringWebTelemetry;
+import java.util.List;
+import org.springframework.http.client.ClientHttpRequestInterceptor;
+import org.springframework.web.client.RestTemplate;
+
+class RestTemplateInterceptor {
```
Comment (laurit) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateInterceptor.java:15]: this class isn't an interceptor but an util to add the interceptor

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java [hunk 6]
```diff
@@ -69,13 +78,15 @@
...
       "otel.logs.exporter=memory"
     })
+@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
 class OtelSpringStarterSmokeTest {
 
   public static final InMemoryMetricExporter METRIC_EXPORTER =
       InMemoryMetricExporter.create(AggregationTemporality.DELTA);
   private static final InMemoryLogRecordExporter LOG_RECORD_EXPORTER =
       InMemoryLogRecordExporter.create();
   public static final InMemorySpanExporter SPAN_EXPORTER = InMemorySpanExporter.create();
```
Comment (laurit) [inline smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:88]: should these also be private?

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java (hunk 4 above)
Comment (zeitlinger) [reply smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:183]: a spring `@Order` annotation is also used in this file

File: instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateInterceptor.java (hunk 5 above)
Comment (jeanbisutti) [inline instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/instrumentation/web/RestTemplateInterceptor.java:15]:
```suggestion
class RestTemplateInstrumentation {
```
?

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java (hunk 6 above)
Comment (jeanbisutti) [inline smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:88]:
```suggestion
  private static final InMemorySpanExporter SPAN_EXPORTER = InMemorySpanExporter.create();
```

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java [hunk 7]
```diff
@@ -157,7 +168,19 @@ AutoConfigurationCustomizerProvider propagatorCustomizer() {
     }
   }
 
+  @AfterEach
+  void tearDown() {
+    reset();
+  }
+
+  private static void reset() {
```
Comment (jeanbisutti) [inline smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:176]:
```suggestion
  private static void resetExporters() {
```

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java [hunk 8]
```diff
@@ -234,4 +255,56 @@ void shouldSendTelemetry() {
         .containsEntry(
             CodeIncubatingAttributes.CODE_NAMESPACE, "org.springframework.boot.StartupInfoLogger");
   }
+
+  @Test
+  @org.junit.jupiter.api.Order(2)
+  void restTemplateClient() {
+    reset(); // ignore the telemetry from application startup
```
Comment (jeanbisutti) [inline smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:262]: `reset()` is already called after each test (see `@AfterEach`). `reset()` call should only be required for the  `restTemplateClient()` test method.  Perhaps remove the `@AfterEach` method?

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java [hunk 9]
```diff
@@ -234,4 +255,56 @@ void shouldSendTelemetry() {
...
+              @Override
+              public void conditionEvaluated(EvaluatedCondition<Object> condition) {}
+
+              @Override
+              public void onTimeout(TimeoutEvent timeoutEvent) {
+                logger.info("Spans: {}", SPAN_EXPORTER.getFinishedSpanItems());
+              }
+            })
+        .await()
+        .atMost(Duration.ofSeconds(5))
```
Comment (jeanbisutti) [inline smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:305]:
```suggestion
        .atMost(Duration.ofSeconds(1))
```

Comment (laurit) [reply smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:305]: @jeanbisutti why do you want to reduce the timeout?

Comment (jeanbisutti) [reply smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:305]: If the test takes more 5 seconds on the CI we could try to modify it. @laurit What is the usual time out for other tests?

Comment (zeitlinger) [reply smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:305]: 1s wasn't always enough I observed

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java (hunk 8 above)
Comment (zeitlinger) [reply smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java]: great observation!

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java (hunk 9 above)
Comment (jeanbisutti) [reply smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:299]: Let's take 5 seconds then

Comment (laurit) [reply smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:299]: > What is the usual time out for other tests?

10 and 30s seem to be the most popular choices. You have to keep in mind that the test suite can generate high load and tests that in isolation pass instantly can take significantly more time. In my opinion it is better to use generous timeouts to ensure that tests won't be flaky because of not waiting long enough.

File: smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java [hunk 10]
```diff
@@ -234,4 +249,56 @@ void shouldSendTelemetry() {
...
```
Comment (zeitlinger) [inline smoke-tests-otel-starter/src/test/java/io/opentelemetry/smoketest/OtelSpringStarterSmokeTest.java:299]:
```suggestion
        .atMost(Duration.ofSeconds(30))
```

**Review by jeanbisutti — APPROVED** [review]
